File: 0064-minimum-path-sum/0064-minimum-path-sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        vector<vector<int>> array(grid.size(), vector<int>(grid[0].size(), 0));
        for (int i = 0; i < 1; i++) {
            // cout << "im herree";
            for (int j = 0; j < grid[i].size(); j++) {
                if (j == 0) {
                    array[i][j] = grid[i][j];
                } else {
                    array[i][j] = grid[i][j] + array[i][j - 1];
                }
            }
        }
        for (int i = 1; i < grid.size(); i++) {
            for (int j = 0; j < 1; j++) {
                array[i][j] = array[i - 1][j] + grid[i][j];
            }
        }
        for (int i = 1; i < grid.size(); i++) {
            for (int j = 1; j < grid[i].size(); j++) {
                array[i][j] = min(grid[i][j] + array[i - 1][j],
                                  grid[i][j] + array[i][j - 1]);
            }
        }
        return array[grid.size() - 1][grid[0].size() - 1];
    }
};

File: 0064-minimum-path-sum/0064-minimum-path-sum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        m=len(grid)-1
        n=len(grid[0])-1
        if m==0 and n ==0:
            return grid[m][n]
        m=len(grid)-1
        n=len(grid[0])-1
        memo={}
        def minimum(m,n):
            if m<0 or n<0:
                return 999999999999999
            if m==0 and n==0:
                return grid[m][n]
            if (m,n) in memo:
                return memo[(m,n)]
            val1=grid[m][n]+minimum(m-1,n)
            val2=grid[m][n]+minimum(m,n-1)
            memo[(m,n)] = min(val1,val2)
            return memo[(m,n)]
        minimum(m,n)
        return memo[(m,n)]